wrench/src/yaml_frame_reader.rs

```diff
@@ -66,23 +94,176 @@ impl YamlFrameReader {
+            }
+            _ => {
+                None
+            }
+        }
+    }
+
     fn handle_rect(&mut self, wrench: &mut Wrench, clip_region: &ClipRegion, item: &Yaml)
     {
-        let rect = item[if item["type"].is_badvalue() { "rect" } else { "bounds" }]
-            .as_rect().expect("rect type must have bounds");
+        let bounds_key = if item["type"].is_badvalue() { "rect" } else { "bounds" };
+        let rect = item[bounds_key].as_rect().expect("rect type must have bounds");
         let color = item["color"].as_colorf().unwrap_or(*WHITE_COLOR);
 
-        let builder = self.builder();
-        let clip = item["clip"].as_clip_region(builder).unwrap_or(*clip_region);
-        builder.push_rect(rect, clip, color);
+        let clip = self.to_clip_region(&item["clip"], &rect, wrench).unwrap_or(*clip_region);
+        self.builder().push_rect(rect, clip, color);
     }
 
-    fn handle_image(&mut self, wrench: &mut Wrench, clip_region: &ClipRegion, item: &Yaml)
+    fn handle_gradient(&mut self, wrench: &mut Wrench, clip_region: &ClipRegion, item: &Yaml)
+    {
+        let bounds_key = if item["type"].is_badvalue() { "gradient" } else { "bounds" };
+        let bounds = item[bounds_key].as_rect().expect("gradient must have bounds");
+        let start = item["start"].as_point().expect("gradient must have start");
+        let end = item["end"].as_point().expect("gradient must have end");
+        let stops = if let Some(stops) = item["stops"].as_vec() {
+            // FIXME(vlad): I can't find the right way to do this with iterators;
+            // I want to take N elements at a time.
+            let num_stops = stops.len()/2;
+            let mut g_stops = Vec::with_capacity(num_stops);
+            for n in 0..num_stops {
+                g_stops.push(GradientStop {
+                    offset: stops[n*2+0].as_force_f32().expect("gradient stop offset is not f32"),
+                    color: stops[n*2+1].as_colorf().expect("gradient stop color is not color"),
+                });
+            }
+            g_stops
+        } else {
+            panic!("gradient must have stops array");
+        };
+
+        let clip = self.to_clip_region(&item["clip"], &bounds, wrench).unwrap_or(*clip_region);
+        self.builder().push_gradient(bounds, clip, start, end, stops);
+    }
+
+    fn handle_border(&mut self, wrench: &mut Wrench, clip_region: &ClipRegion, item: &Yaml)
+    {
+        let bounds_key = if item["type"].is_badvalue() { "border" } else { "bounds" };
+        let bounds = item[bounds_key].as_rect().expect("borders must have bounds");
+        let widths = item["width"].as_vec_f32().expect("borders must have width(s)");
+        let colors = item["color"].as_vec_colorf().expect("borders must have color(s)");
+        let styles = item["style"].as_vec_string().expect("borders must have style(s)");
+        let styles = styles.iter().map(|s| match s {
+            s if s == "none" => BorderStyle::None,
```

Comment:
You should be able to just do `"something" => BorderStyle::Something` without all the `if` in the matching